### ISSUE TITLE
moe routing loss incorrect for post-training

### ISSUE THREAD

**Issue by gshennvm** [NVIDIA/Megatron-LM#1984]
**Describe the bug**

The moe router loss assumes that all tokens should be counted in the loss https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/moe/moe_utils.py#L108, however this seems incorrect in post-training when we have padding tokens(the padding tokens would count towards the routing loss). There is also another thing to be said if we want to include routing loss for the prompt or just the generations in RL.

**Expected behavior**

perhaps moe routing loss should take a mask of the token routing, so we can avoid computing that in the loss.

**Comment by yanring** [comment]
Thanks for reporting this issue. We are already aware of this issue and working on it.

**Comment by Victarry** [comment]
In my mind, there are several codes changes needed to fix this issue:
1. Pass the mask of padding tokens to the MoE layer and Router.
2. Calculate the correct tokens_per_expert for aux_loss considering padding mask.
3. Prevent the backward gradient flow to the logits of padded tokens.
4. Work with calculate_per_token_loss=True with padded token mask.
5. Add an UT to guard such cases.
+@HaochenYuan
